src/main/java/io/fabric8/elasticsearch/plugin/KibanaUserReindexFilter.java

```diff
@@ -0,0 +1,97 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.elasticsearch.plugin;
+
+import org.apache.commons.lang.ObjectUtils;
+import org.apache.commons.lang.StringUtils;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.netty.handler.codec.http.HttpRequest;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.http.netty.NettyHttpRequest;
+import org.elasticsearch.rest.RestChannel;
+import org.elasticsearch.rest.RestFilter;
+import org.elasticsearch.rest.RestFilterChain;
+import org.elasticsearch.rest.RestRequest;
+
+public class KibanaUserReindexFilter extends RestFilter {
+
+	private static final String DEFAULT_AUTH_PROXY_HEADER = "X-Proxy-Remote-User";
+	private static final String SEARCHGUARD_AUTHENTICATION_PROXY_HEADER = "searchguard.authentication.proxy.header";
+
+	private final ESLogger logger;
+	private final String proxyUserHeader;
+
+	public KibanaUserReindexFilter(final Settings settings, final ESLogger logger){
+		this.logger = logger;
+		this.proxyUserHeader = settings.get(SEARCHGUARD_AUTHENTICATION_PROXY_HEADER, DEFAULT_AUTH_PROXY_HEADER);
+	}
+	
+	@Override
+	public void process(RestRequest request, RestChannel channel, RestFilterChain chain) throws Exception {
+		try {
+			logger.debug("Handling Request in Kibana User Reindex filter...");
+			final String user = getUser(request);
+			final String requestedIndex = getRequestedIndex(request);
+			
+			// rewrite .kibana indexes to .kibana.<username> based on the user header 
+			if ( StringUtils.isNotEmpty(user) && StringUtils.isNotEmpty(requestedIndex) && requestedIndex.equalsIgnoreCase(".kibana")) {
+				String userIndex = requestedIndex + "." + user;
+				logger.debug("Matched for a kibana user header, will set to '{}' for user '{}'", userIndex, user);
+				
+				//update the request URI here
+				request = updateRequestIndex(request, requestedIndex, userIndex);
+				
+				logger.debug("URI for request is '{}' after update", request.uri());
+			}
+
+		} catch (Exception e) {
+			logger.error("Error handling request in OpenShift SearchGuard filter", e);
+		} finally {
+			chain.continueProcessing(request, channel);
+		}
+	}
+
+	private String getUser(RestRequest request) {
+		return (String) ObjectUtils.defaultIfNull(request.header(proxyUserHeader), "");
+	}
+	
+	private RestRequest updateRequestIndex(RestRequest request, String oldIndex, String newIndex) {
+		String uri = request.uri();
+		String replaced = uri.replace(oldIndex, newIndex);
+		
+		NettyHttpRequest nettyRequest = (NettyHttpRequest) request;
+		HttpRequest httpRequest = nettyRequest.request();
+		httpRequest.setUri(replaced);
+		
+		return new NettyHttpRequest(httpRequest, nettyRequest.getChannel());
+	}
+	
+	private String getRequestedIndex(RestRequest request) {
```

Comment:
this is a little misleading I think, request.uri is just a string, and if we used the java.net.uri package we would end up creating a new URI object and parsing it to get the path which is what we already get from request.uri...